File: src/coordinate-canonicalization.ts
/**
 * Stable, allow-listed persisted geometry.
 *
 * Mirrored by custom_components/houseplan/coordinate_canonicalization.py.
 * Keep the precision, lattice formula and field allow-list in lockstep.
 */

import {
  DECOR_BOX_KINDS,
  type DecorBoxKind,
} from './editors/decor/types';

export { DECOR_BOX_KINDS };

export const COORDINATE_DECIMALS = 9;
export const COORDINATE_FACTOR = 10 ** COORDINATE_DECIMALS;
export const LATTICE_GRID_N = 240;
export const LATTICE_NOISE_STEPS = 1e-4;

type JsonRecord = Record<string, any>;

export interface LatticeSpaceReport {
  spaceId: string;
  space: string;
  canonicalized: number;
  far: number;
  maxShift: number;
  maxShiftCm: number;
}

export interface LatticeCanonicalizationReport {
  canonicalized: number;
  far: number;
  maxShift: number;
  maxShiftCm: number;
  spaces: LatticeSpaceReport[];
}

interface MutableLatticeReport extends LatticeCanonicalizationReport {
  bySpace: Map<string, LatticeSpaceReport>;
}

function cloneJson<T>(value: T): T {
  if (Array.isArray(value)) return value.map((item) => cloneJson(item)) as T;
  if (value !== null && typeof value === 'object') {
    const out: JsonRecord = {};
    for (const [key, item] of Object.entries(value as JsonRecord)) out[key] = cloneJson(item);
    return out as T;
  }
  return value;
}

/** Existing scalar contract for transforms, angles, lengths and ratios. */
export function canonicalizeNumber(value: unknown): unknown {
  if (typeof value !== 'number' || !Number.isFinite(value)) return value;
  const sign = value < 0 || Object.is(value, -0) ? -1 : 1;
  const result = sign
    * (Math.floor(Math.abs(value) * COORDINATE_FACTOR + 0.5) / COORDINATE_FACTOR);
  return result === 0 ? 0 : result;
}

/** Collapse only an unobservable tail around a 1/240 node. */
export function canonicalizeLatticeCoordinate(value: unknown): unknown {
  if (typeof value !== 'number' || !Number.isFinite(value)) return value;
  const scaled = value * LATTICE_GRID_N;
  const nearest = Math.round(scaled);
  const deviation = Math.abs(scaled - nearest);
  if (deviation < LATTICE_NOISE_STEPS) {
    const result = nearest / LATTICE_GRID_N;
    return result === 0 ? 0 : result;
  }
  return canonicalizeNumber(value);
}

/** Three significant digits without turning a non-zero sub-millimetre shift into zero. */
export function formatLatticeShiftCm(value: number): string {
  if (!Number.isFinite(value) || value === 0) return '0';
  const absolute = Math.abs(value);
  if (absolute < 0.001) return value.toExponential(2);
  return String(Number(value.toPrecision(3)));
}

function record(value: unknown): JsonRecord | null {
  return value !== null && typeof value === 'object' && !Array.isArray(value)
    ? value as JsonRecord : null;
}

function records(value: unknown): JsonRecord[] {
  return Array.isArray(value)
    ? value.filter((item): item is JsonRecord => record(item) !== null)
    : [];
}

function isDecorBoxKind(value: unknown): value is DecorBoxKind {
  return typeof value === 'string'
    && (DECOR_BOX_KINDS as readonly string[]).includes(value);
}

function scalarFields(item: JsonRecord, names: readonly string[]): void {
  for (const name of names) {
    if (Object.prototype.hasOwnProperty.call(item, name)) {
      item[name] = canonicalizeNumber(item[name]);
    }
  }
}

function latticeFields(item: JsonRecord, names: readonly string[]): void {
  for (const name of names) {
    if (Object.prototype.hasOwnProperty.call(item, name)) {
      item[name] = canonicalizeLatticeCoordinate(item[name]);
    }
  }
}

function latticePoint(value: unknown): void {
  if (!Array.isArray(value)) return;
  for (let index = 0; index < Math.min(2, value.length); index++) {
    value[index] = canonicalizeLatticeCoordinate(value[index]);
  }
}

function latticePoints(value: unknown): void {
  if (!Array.isArray(value)) return;
  for (const item of value) latticePoint(item);
}

/** Visit exactly the persisted coordinate allow-list, without cloning. */
function visitLatticeCoordinates(
  config: unknown,
  layout: unknown,
  visit: (value: number, space: JsonRecord | null) => void,
): void {
  const root = record(config);
  if (root) {
    for (const space of records(root.spaces)) {
      const fieldValues = (item: JsonRecord, names: readonly string[]): void => {
        for (const name of names) {
          const value = item[name];
          if (typeof value === 'number' && Number.isFinite(value)) visit(value, space);
        }
      };
      const pointValues = (value: unknown): void => {
        if (!Array.isArray(value)) return;
        for (let index = 0; index < Math.min(2, value.length); index++) {
          const coordinate = value[index];
          if (typeof coordinate === 'number' && Number.isFinite(coordinate)) {
            visit(coordinate, space);
          }
        }
      };
      const pointsValues = (value: unknown): void => {
        if (Array.isArray(value)) for (const point of value) pointValues(point);
      };

      for (const room of records(space.rooms)) {
        fieldValues(room, ['x', 'y', 'w', 'h']);
        pointsValues(room.poly);
      }
      for (const wall of records(space.walls)) {
        pointValues(wall.a);
        pointValues(wall.b);
      }
      for (const segment of records(space.wall_segments)) {
        pointValues(segment.a);
        pointValues(segment.b);
      }
      for (const opening of records(space.openings)) fieldValues(opening, ['x', 'y']);
      for (const decor of records(space.decor)) {
        if (decor.kind === 'line') fieldValues(decor, ['x1', 'y1', 'x2', 'y2']);
        else if (isDecorBoxKind(decor.kind)) {
          fieldValues(decor, ['x', 'y', 'w', 'h']);
        } else if (decor.kind === 'text') fieldValues(decor, ['x', 'y']);
      }
      for (const draft of records(space.room_drafts)) pointsValues(draft.points);
      for (const partition of records(space.partitions)) {
        pointValues(partition.a);
        pointValues(partition.b);
      }
      for (const column of records(space.wall_columns)) pointValues(column.center);
      for (const span of records(space.open_spans)) {
        pointValues(span.a);
        pointValues(span.b);
      }
    }
  }

  const layoutRoot = record(layout);
  if (!layoutRoot) return;
  const spacesById = new Map<string, JsonRecord>();
  if (root) {
    for (const space of records(root.spaces)) {
      if (space.id != null) spacesById.set(String(space.id), space);
    }
  }
  for (const value of Object.values(layoutRoot)) {
    const item = record(value);
    if (!item) continue;
    const owner = item.s != null ? spacesById.get(String(item.s)) || null : null;
    for (const name of ['x', 'y']) {
      const coordinate = item[name];
      if (typeof coordinate === 'number' && Number.isFinite(coordinate)) visit(coordinate, owner);
    }
  }
}

function emptyReport(): MutableLatticeReport {
  return {
    canonicalized: 0,
    far: 0,
    maxShift: 0,
    maxShiftCm: 0,
    spaces: [],
    bySpace: new Map(),
  };
}

function cellCm(space: JsonRecord | null): number {
  const value = Number(space?.cell_cm);
  return value > 0 ? value : 5;
}

/** Measure the exact work the boundary will do, without cloning or writing. */
export function latticeCanonicalizationReport(
  config: unknown,
  layout: unknown = {},
): LatticeCanonicalizationReport {
  const report = emptyReport();
  const spaces = records(record(config)?.spaces);
  let worstCellCm = 5;
  for (const space of spaces) worstCellCm = Math.max(worstCellCm, cellCm(space));

  visitLatticeCoordinates(config, layout, (value, space) => {
    const scaled = value * LATTICE_GRID_N;
    const deviation = Math.abs(scaled - Math.round(scaled));
    const canonical = canonicalizeLatticeCoordinate(value);
    const isNoise = deviation > 0 && deviation < LATTICE_NOISE_STEPS;
    const isFar = deviation >= LATTICE_NOISE_STEPS;
    if (!isNoise && !isFar) return;

    const shift = isNoise && typeof canonical === 'number' ? Math.abs(canonical - value) : 0;
    if (isNoise) report.canonicalized++;
    else report.far++;
    report.maxShift = Math.max(report.maxShift, shift);
    report.maxShiftCm = Math.max(
      report.maxShiftCm,
      shift * LATTICE_GRID_N * (space ? cellCm(space) : worstCellCm),
    );

    if (!space?.id || !isNoise) return;
    const id = String(space.id);
    let item = report.bySpace.get(id);
    if (!item) {
      item = {
        spaceId: id,
        space: String(space.title || id),
        canonicalized: 0,
        far: 0,
        maxShift: 0,
        maxShiftCm: 0,
      };
      report.bySpace.set(id, item);
    }
    item.canonicalized++;
    item.maxShift = Math.max(item.maxShift, shift);
    item.maxShiftCm = Math.max(item.maxShiftCm, shift * LATTICE_GRID_N * cellCm(space));
  });

  // Far values are useful only next to a space that the barrier actually
  // touched. Count them in a second allocation-free pass into those rows.
  if (report.bySpace.size) {
    visitLatticeCoordinates(config, layout, (value, space) => {
      if (!space?.id) return;
      const item = report.bySpace.get(String(space.id));
      if (!item) return;
      const scaled = value * LATTICE_GRID_N;
      if (Math.abs(scaled - Math.round(scaled)) >= LATTICE_NOISE_STEPS) item.far++;
    });
  }
  report.spaces = [...report.bySpace.values()];
  const { bySpace: _bySpace, ...publicReport } = report;
  return publicReport;
}

export function canonicalizePosition<T>(position: T): T {
  const result = cloneJson(position);
  return canonicalizePositionInPlace(result);
}

export function canonicalizePositionInPlace<T>(position: T): T {
  const result = position;
  const item = record(result);
  if (item) latticeFields(item, ['x', 'y']);
  return result;
}

export function canonicalizeLayoutGeometry<T>(layout: T): T {
  const result = cloneJson(layout);
  return canonicalizeLayoutGeometryInPlace(result);
}

export function canonicalizeLayoutGeometryInPlace<T>(layout: T): T {
  const result = layout;
  const root = record(result);
  if (!root) return result;
  for (const value of Object.values(root)) {
    const item = record(value);
    if (item) latticeFields(item, ['x', 'y']);
  }
  return result;
}

export function canonicalizeConfigGeometry<T>(config: T): T {
  const result = cloneJson(config);
  return canonicalizeConfigGeometryInPlace(result);
}

export function canonicalizeConfigGeometryInPlace<T>(config: T): T {
  const result = config;
  const root = record(result);
  if (!root) return result;

  for (const space of records(root.spaces)) {
    scalarFields(space, [
      'plan_x', 'plan_y', 'plan_scale', 'plan_scale_x', 'plan_scale_y', 'plan_angle',
    ]);

    for (const room of records(space.rooms)) {
      latticeFields(room, ['x', 'y', 'w', 'h']);
      latticePoints(room.poly);
    }

    for (const wall of records(space.walls)) {
      latticePoint(wall.a);
      latticePoint(wall.b);
    }
    for (const segment of records(space.wall_segments)) {
      latticePoint(segment.a);
      latticePoint(segment.b);
    }

    for (const opening of records(space.openings)) {
      latticeFields(opening, ['x', 'y']);
      scalarFields(opening, ['angle', 'length']);
      const host = record(opening.host);
      if (host) scalarFields(host, ['t']);
    }

    for (const decor of records(space.decor)) {
      if (decor.kind === 'line') latticeFields(decor, ['x1', 'y1', 'x2', 'y2']);
      else if (isDecorBoxKind(decor.kind)) {
        latticeFields(decor, ['x', 'y', 'w', 'h']);
        scalarFields(decor, ['angle']);
      } else if (decor.kind === 'text') {
        latticeFields(decor, ['x', 'y']);
        scalarFields(decor, ['scale', 'angle']);
      }
    }

    for (const draft of records(space.room_drafts)) latticePoints(draft.points);

    for (const partition of records(space.partitions)) {
      latticePoint(partition.a);
      latticePoint(partition.b);
    }

    for (const column of records(space.wall_columns)) {
      latticePoint(column.center);
      if (column.shape === 'square') scalarFields(column, ['angle']);
    }

    for (const span of records(space.open_spans)) {
      latticePoint(span.a);
      latticePoint(span.b);
    }
  }

  for (const marker of records(root.markers)) scalarFields(marker, ['angle']);
  return result;
}

File: custom_components/houseplan/coordinate_canonicalization.py
"""Lossless, allow-listed canonicalisation of persisted geometry.

The frontend mirrors this module in src/coordinate-canonicalization.ts.
Keep the precision, lattice formula and field allow-list in lockstep; a shared
fixture is exercised by both runtimes.
"""
from __future__ import annotations

import copy
import math
from typing import Any


COORDINATE_DECIMALS = 9
COORDINATE_FACTOR = 10**COORDINATE_DECIMALS
LATTICE_GRID_N = 240
LATTICE_NOISE_STEPS = 1e-4


def canonicalize_number(value: Any) -> Any:
    """Return one stable IEEE-754 representation for an allow-listed scalar."""
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        return value
    number = float(value)
    if not math.isfinite(number):
        return value
    sign = -1.0 if math.copysign(1.0, number) < 0 else 1.0
    result = sign * (
        math.floor(abs(number) * COORDINATE_FACTOR + 0.5)
        / COORDINATE_FACTOR
    )
    if result == 0:
        return 0.0
    return result


def canonicalize_lattice_coordinate(value: Any) -> Any:
    """Collapse near-node noise while preserving authored off-grid values."""
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        return value
    number = float(value)
    if not math.isfinite(number):
        return value
    scaled = number * LATTICE_GRID_N
    # JavaScript Math.round: ties go toward +infinity, unlike Python round().
    nearest = math.floor(scaled + 0.5)
    if abs(scaled - nearest) < LATTICE_NOISE_STEPS:
        result = nearest / LATTICE_GRID_N
        return 0.0 if result == 0 else result
    return canonicalize_number(number)


def _record(value: Any) -> dict[str, Any] | None:
    return value if isinstance(value, dict) else None


def _records(value: Any) -> list[dict[str, Any]]:
    if not isinstance(value, list):
        return []
    return [item for item in value if isinstance(item, dict)]


def _scalar_fields(record: dict[str, Any], names: tuple[str, ...]) -> None:
    for name in names:
        if name in record:
            record[name] = canonicalize_number(record[name])


def _lattice_fields(record: dict[str, Any], names: tuple[str, ...]) -> None:
    for name in names:
        if name in record:
            record[name] = canonicalize_lattice_coordinate(record[name])


def _lattice_point(value: Any) -> None:
    if not isinstance(value, list):
        return
    for index in range(min(2, len(value))):
        value[index] = canonicalize_lattice_coordinate(value[index])


def _lattice_points(value: Any) -> None:
    if not isinstance(value, list):
        return
    for point in value:
        _lattice_point(point)


def canonicalize_position(position: Any) -> Any:
    """Canonicalise lattice x/y in one layout record, preserving metadata."""
    result = copy.deepcopy(position)
    record = _record(result)
    if record is not None:
        _lattice_fields(record, ("x", "y"))
    return result


def canonicalize_layout_geometry(layout: Any) -> Any:
    """Canonicalise lattice x/y in every layout record."""
    result = copy.deepcopy(layout)
    record = _record(result)
    if record is None:
        return result
    for position in record.values():
        item = _record(position)
        if item is not None:
            _lattice_fields(item, ("x", "y"))
    return result


def canonicalize_config_geometry(config: Any) -> Any:
    """Canonicalise only the named persisted geometry fields."""
    result = copy.deepcopy(config)
    root = _record(result)
    if root is None:
        return result

    for space in _records(root.get("spaces")):
        _scalar_fields(
            space,
            (
                "plan_x",
                "plan_y",
                "plan_scale",
                "plan_scale_x",
                "plan_scale_y",
                "plan_angle",
            ),
        )

        for room in _records(space.get("rooms")):
            _lattice_fields(room, ("x", "y", "w", "h"))
            _lattice_points(room.get("poly"))

        for wall in _records(space.get("walls")):
            _lattice_point(wall.get("a"))
            _lattice_point(wall.get("b"))

        for opening in _records(space.get("openings")):
            _lattice_fields(opening, ("x", "y"))
            _scalar_fields(opening, ("angle", "length"))
            host = _record(opening.get("host"))
            if host is not None:
                _scalar_fields(host, ("t",))

        for decor in _records(space.get("decor")):
            kind = decor.get("kind")
            if kind == "line":
                _lattice_fields(decor, ("x1", "y1", "x2", "y2"))
            elif kind in ("rect", "ellipse", "furniture"):
                _lattice_fields(decor, ("x", "y", "w", "h"))
                _scalar_fields(decor, ("angle",))
            elif kind == "text":
                _lattice_fields(decor, ("x", "y"))
                _scalar_fields(decor, ("scale", "angle"))

        for draft in _records(space.get("room_drafts")):
            _lattice_points(draft.get("points"))

        for partition in _records(space.get("partitions")):
            _lattice_point(partition.get("a"))
            _lattice_point(partition.get("b"))

        for column in _records(space.get("wall_columns")):
            _lattice_point(column.get("center"))
            if column.get("shape") == "square":
                _scalar_fields(column, ("angle",))

        for span in _records(space.get("open_spans")):
            _lattice_point(span.get("a"))
            _lattice_point(span.get("b"))

    for marker in _records(root.get("markers")):
        _scalar_fields(marker, ("angle",))

    return result
